File: aggregate.py
#!/usr/bin/env python3
"""
aggregate.py — aggregate hyperfine benchmark results for jvm-build-tools-bench

Expected input layout (produced by run_bench.sh):
  <results-dir>/
    <repo>/
      <tool>-<scenario>.json   (hyperfine JSON export)

Repos and tools are discovered dynamically from the directory structure and
filenames. Scenario names are matched against a small canonical set so
multi-word names like `clean-compile` stay stable.

Outputs (written to --output-dir/<repo>/<scenario>/):
  summary.json  structured data: mean/stddev/min/max per tool for one scenario
  summary.md    markdown comparison table for one scenario
  chart.svg     horizontal bar chart for one scenario (requires matplotlib)

Also writes --output-dir/index.html: per-repo collapsible sections with lazy-loaded SVGs.

ASCII bar charts are also printed to stdout (useful in CI logs, no extra deps).

Usage:
  python aggregate.py --results-dir results/ [--output-dir aggregated/]
"""

import argparse
import datetime
import json
import sys
from collections import defaultdict
from pathlib import Path

from tool_colors import get_tool_color

SCENARIO_ALIASES = {
    "clean-compile": "clean-compile",
    "incremental-compile": "incremental-compile",
    "clean": "clean-compile",
    "incremental": "incremental-compile",
}


# ── CLI ──────────────────────────────────────────────────────────────────────

def parse_args():
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("--results-dir", required=True,
                   help="Directory with <repo>/<tool>-<scenario>.json files")
    p.add_argument("--output-dir", default="aggregated",
                   help="Where to write output files (default: aggregated/)")
    return p.parse_args()


# ── Data loading ──────────────────────────────────────────────────────────────

def parse_result_stem(stem: str) -> tuple[str, str] | None:
    for suffix, scenario in sorted(
        SCENARIO_ALIASES.items(), key=lambda entry: len(entry[0]), reverse=True
    ):
        needle = f"-{suffix}"
        if stem.endswith(needle):
            tool = stem[: -len(needle)]
            if tool:
                return tool, scenario
    return None

def load_results(results_dir: Path) -> dict[str, dict[str, list[dict]]]:
    """
    Walk results_dir and parse every <repo>/<tool>-<scenario>.json.
    Returns { repo: { scenario: [ {tool, mean, stddev, min, max, median}, ... ] } }
    sorted by mean within each scenario.
    Repo is the immediate parent directory name of each JSON file.
    """
    # { repo: { scenario: { tool: entry } } }
    by_repo: dict[str, dict[str, dict[str, dict]]] = defaultdict(lambda: defaultdict(dict))

    for path in sorted(results_dir.rglob("*.json")):
        repo = path.parent.name   # immediate parent dir = repo name
        stem = path.stem          # e.g. "maven-clean"

        parsed = parse_result_stem(stem)
        if parsed is None:
            print(f"  Skipping {path} (unknown scenario suffix)", file=sys.stderr)
            continue

        tool, scenario = parsed
        get_tool_color(tool)

        try:
            data = json.loads(path.read_text())
        except (json.JSONDecodeError, OSError) as e:
            print(f"  Warning: could not read {path}: {e}", file=sys.stderr)
            continue

        results = data.get("results", [])
        if not results:
            print(f"  Warning: no results in {path}", file=sys.stderr)
            continue

        r = results[0]
        by_repo[repo][scenario][tool] = {
            "tool":   tool,
            "mean":   r.get("mean", 0.0),
            "stddev": r.get("stddev", 0.0),
            "min":    r.get("min", 0.0),
            "max":    r.get("max", 0.0),
            "median": r.get("median", r.get("mean", 0.0)),
        }

    # Sort each scenario's entries by mean (fastest first)
    return {
        repo: {
            scenario: sorted(entries.values(), key=lambda e: e["mean"])
            for scenario, entries in sorted(scenarios.items())
        }
        for repo, scenarios in sorted(by_repo.items())
    }


# ── JSON summary ──────────────────────────────────────────────────────────────

def write_summary_json(repo: str, scenario: str, entries: list[dict], output_dir: Path):
    out = output_dir / "summary.json"
    out.write_text(json.dumps({
        "repo": repo,
        "scenario": scenario,
        "entries": entries,
    }, indent=2))
    print(f"  Wrote {out}")


# ── Markdown table ────────────────────────────────────────────────────────────

def fmt_ms(seconds: float) -> str:
    return f"{seconds * 1000:.1f} ms"

def write_summary_md(repo: str, scenario: str, entries: list[dict], output_dir: Path):
    lines = [f"# JVM Build Tools Benchmark — {repo} — {scenario.replace('-', ' ').title()}\n"]
    lines.append("| Tool | Mean | Stddev | Min | Max |")
    lines.append("|------|-----:|-------:|----:|----:|")
    for e in entries:
        lines.append(
            f"| {e['tool']} "
            f"| {fmt_ms(e['mean'])} "
            f"| ± {fmt_ms(e['stddev'])} "
            f"| {fmt_ms(e['min'])} "
            f"| {fmt_ms(e['max'])} |"
        )
    lines.append("")

    out = output_dir / "summary.md"
    out.write_text("\n".join(lines))
    print(f"  Wrote {out}")


# ── ASCII bar chart ───────────────────────────────────────────────────────────

BAR_WIDTH = 40
BAR_CHAR = "█"

def ascii_bar_chart(repo: str, scenario: str, entries: list[dict]):
    if not entries:
        return

    max_mean = max(e["mean"] for e in entries)
    max_tool_len = max(len(e["tool"]) for e in entries)

    print(f"\n  [{repo}] {scenario} (mean time, lower is better)")
    for e in entries:
        bar_len = int(round((e["mean"] / max_mean) * BAR_WIDTH)) if max_mean > 0 else 0
        bar = BAR_CHAR * bar_len
        label = e["tool"].ljust(max_tool_len)
        mean_ms = e["mean"] * 1000
        stddev_ms = e["stddev"] * 1000
        print(f"    {label}  {bar:<{BAR_WIDTH}}  {mean_ms:8.1f} ms ± {stddev_ms:.1f}")


# ── SVG / matplotlib chart ────────────────────────────────────────────────────

def write_svg_chart(repo: str, scenario: str, entries: list[dict], output_dir: Path):
    try:
        import matplotlib
        matplotlib.use("Agg")
        import matplotlib.pyplot as plt
    except ImportError:
        print(f"  matplotlib not available — skipping SVG chart for {scenario}", file=sys.stderr)
        return

    tools = list(reversed([e["tool"] for e in entries]))
    means = list(reversed([e["mean"] * 1000 for e in entries]))
    errors = list(reversed([e["stddev"] * 1000 for e in entries]))

    fig_height = max(3, len(tools) * 0.6 + 1.2)
    fig, ax = plt.subplots(figsize=(10, fig_height))

    bars = ax.barh(
        tools, means,
        xerr=errors,
        capsize=4,
        color=[get_tool_color(tool) for tool in tools],
        edgecolor="white",
        linewidth=0.5,
        error_kw={"elinewidth": 1.2, "ecolor": "#555"},
    )

    for bar, mean in zip(bars, means):
        ax.text(
            bar.get_width() + max(means) * 0.01,
            bar.get_y() + bar.get_height() / 2,
            f"{mean:.1f} ms",
            va="center", ha="left", fontsize=9,
        )

    ax.set_xlabel("Mean time (ms) — lower is better")
    ax.set_title(f"{repo}: {scenario.replace('-', ' ').title()}")
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.set_xlim(0, max(means) * 1.2)

    plt.tight_layout()
    out = output_dir / "chart.svg"
    fig.savefig(out, format="svg", bbox_inches="tight")
    plt.close(fig)
    print(f"  Wrote {out}")


# ── HTML index ───────────────────────────────────────────────────────────────

_HTML_CSS = """
  body { font-family: system-ui, sans-serif; max-width: 960px; margin: 2rem auto; padding: 0 1rem; color: #222; }
  h1 { font-size: 1.6rem; }
  details { border: 1px solid #ddd; border-radius: 6px; margin: 1rem 0; }
  summary { cursor: pointer; padding: 0.75rem 1rem; font-size: 1.1rem; font-weight: 600;
            background: #f5f5f5; border-radius: 6px; list-style: none; }
  summary::-webkit-details-marker { display: none; }
  summary::before { content: "▶ "; font-size: 0.8em; color: #666; }
  details[open] summary::before { content: "▼ "; }
  .repo-body { padding: 1rem; }
  .charts { display: flex; flex-wrap: wrap; gap: 1rem; }
  figure { margin: 0; flex: 1 1 400px; }
  figure img { width: 100%; height: auto; border: 1px solid #eee; border-radius: 4px; }
  figcaption { font-size: 0.85rem; color: #555; margin-top: 0.3rem; text-align: center; }
  .links { margin-top: 0.75rem; font-size: 0.9rem; }
  .links a { margin-right: 1rem; color: #0066cc; }
  footer { margin-top: 2rem; font-size: 0.8rem; color: #999; }
"""

def write_index_html(all_data: dict, output_dir: Path):
    timestamp = datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%d %H:%M UTC")
    sections = []

    for i, (repo, data) in enumerate(sorted(all_data.items())):
        open_attr = " open" if i == 0 else ""
        charts_html = "\n".join(
            f'      <figure>\n'
            f'        <img src="{repo}/{scenario}/chart.svg" alt="{scenario} chart" loading="lazy">\n'
            f'        <figcaption>{scenario.replace("-", " ").title()} '
            f'(<a href="{repo}/{scenario}/summary.json">json</a>, '
            f'<a href="{repo}/{scenario}/summary.md">md</a>)</figcaption>\n'
            f'      </figure>'
            for scenario in sorted(data.keys())
        )
        sections.append(f"""\
  <details{open_attr}>
    <summary>{repo}</summary>
    <div class="repo-body">
      <div class="charts">
{charts_html}
      </div>
    </div>
  </details>""")

    html = f"""\
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>JVM Build Tools Benchmark</title>
  <style>{_HTML_CSS}  </style>
</head>
<body>
  <h1>JVM Build Tools Benchmark</h1>
  <p>Compilation time comparison across build tools on real-world repos.
     Each chart shows mean time (lower is better).</p>

{"".join(chr(10) + s for s in sections)}

  <footer>Generated {timestamp}</footer>
</body>
</html>
"""
    out = output_dir / "index.html"
    out.write_text(html)
    print(f"  Wrote {out}")


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    args = parse_args()
    results_dir = Path(args.results_dir)
    output_dir = Path(args.output_dir)

    if not results_dir.exists():
        print(f"Error: results directory not found: {results_dir}", file=sys.stderr)
        sys.exit(1)

    print(f"Scanning {results_dir} for benchmark results...")
    try:
        all_data = load_results(results_dir)
    except ValueError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

    if not all_data:
        print("No benchmark results found.", file=sys.stderr)
        sys.exit(1)

    for repo, data in all_data.items():
        print(f"\n=== {repo} ===")
        for scenario, entries in data.items():
            scenario_out = output_dir / repo / scenario
            scenario_out.mkdir(parents=True, exist_ok=True)
            write_summary_json(repo, scenario, entries, scenario_out)
            write_summary_md(repo, scenario, entries, scenario_out)
            ascii_bar_chart(repo, scenario, entries)
            write_svg_chart(repo, scenario, entries, scenario_out)

    output_dir.mkdir(parents=True, exist_ok=True)
    write_index_html(all_data, output_dir)
    print("\nDone.")


if __name__ == "__main__":
    main()

File: aggregate.scala
//> using file bench.scala

package bench

import mainargs.{main, arg, ParserForClass}
import java.awt.Color
import java.time.{ZoneOffset, ZonedDateTime}
import java.time.format.DateTimeFormatter

@main
case class Aggregate(
  @arg(name = "results-dir", doc = "Directory with <repo>/<tool>-<scenario>.json files")
  resultsDir: String,
  @arg(name = "output-dir", doc = "Where to write aggregated outputs")
  outputDir: String = "aggregated",
  @arg(name = "benchmarks-yaml", doc = "Path to benchmarks.yaml to discover valid benchmark types")
  benchmarksYaml: String = "benchmarks.yaml"
)

object Aggregate:
  def main(args: Array[String]): Unit =
    val config = ParserForClass[Aggregate].constructEither(args.toSeq) match
      case Left(errMsg) =>
        System.err.println(errMsg)
        sys.exit(1)
      case Right(c) => c

    val resultsPath = os.Path(config.resultsDir, os.pwd)
    val outputPath = os.Path(config.outputDir, os.pwd)

    val benchmarksYamlPath = os.Path(config.benchmarksYaml, os.pwd)
    val benchConfig = Config.load(benchmarksYamlPath)
    val validTypes = benchConfig.hyperfine.benchmark_types.keySet
    println(s"Valid benchmark types: ${validTypes.toSeq.sorted.mkString(", ")}")

    if !os.exists(resultsPath) then
      System.err.println(s"Error: results directory not found: $resultsPath")
      sys.exit(1)

    println(s"Scanning $resultsPath for benchmark results...")
    val allData = loadResults(resultsPath, validTypes)
    if allData.isEmpty then
      println("No benchmark results found.")
      sys.exit(1)

    for (repo, scenarios) <- allData.toSeq.sortBy(_._1) do
      println(s"\n=== $repo ===")
      for (scenario, entries) <- scenarios do
        val scenarioOut = outputPath / repo / scenario
        os.makeDir.all(scenarioOut)

        writeSummaryJson(repo, scenario, entries, scenarioOut)
        writeSummaryMd(repo, scenario, entries, scenarioOut)
        asciiBarChart(repo, scenario, entries)
        writeSvgChart(repo, scenario, entries, scenarioOut)

    os.makeDir.all(outputPath)
    writeIndexHtml(allData, outputPath)
    println("\nDone.")

  // ── Data loading ──────────────────────────────────────────────────────────

  private def loadResults(resultsDir: os.Path, validTypes: Set[String]): Map[String, Map[String, List[BenchmarkEntry]]] =
    import scala.collection.mutable

    val byRepo = mutable.Map.empty[String, mutable.Map[String, mutable.Map[String, BenchmarkEntry]]]

    for path <- os.walk(resultsDir) if path.ext == "json" && !path.last.startsWith(".") do
      val repo = if path / os.up == resultsDir then resultsDir.last else (path / os.up).last

      val stem = path.last.stripSuffix(".json")
      parseResultStem(stem, validTypes) match
        case Some((tool, scenario)) =>
          // Validate tool color knowledge
          try ToolColors.get(tool)
          catch
            case e: IllegalArgumentException =>
              System.err.println(s"Error: ${e.getMessage}")
              sys.exit(1)

          val parsed = tryParseHyperfine(path)
          parsed.foreach { entry =>
            val e = entry.copy(tool = tool)
            val repoMap = byRepo.getOrElseUpdate(repo, mutable.Map.empty)
            val scenarioMap = repoMap.getOrElseUpdate(scenario, mutable.Map.empty)
            scenarioMap(tool) = e
          }
        case None =>
          System.err.println(s"  Skipping $path (unknown scenario suffix)")

    // Convert to immutable, sorted by mean (fastest first)
    byRepo.view.mapValues { scenarios =>
      scenarios.view.mapValues { entries =>
        entries.values.toList.sortBy(_.mean)
      }.toMap
    }.toMap

  private def tryParseHyperfine(path: os.Path): Option[BenchmarkEntry] =
    try
      val content = os.read(path)
      val data = upickle.default.read[Map[String, ujson.Value]](content)
      val results = data.getOrElse("results", ujson.Arr()).arr
      if results.isEmpty then
        System.err.println(s"  Warning: no results in $path")
        None
      else
        val r = results.head.obj
        Some(BenchmarkEntry(
          tool = "",
          mean = r.getOrElse("mean", ujson.Num(0.0)).num,
          stddev = r.getOrElse("stddev", ujson.Num(0.0)).num,
          min = r.getOrElse("min", ujson.Num(0.0)).num,
          max = r.getOrElse("max", ujson.Num(0.0)).num,
          median = r.getOrElse("median", r.getOrElse("mean", ujson.Num(0.0))).num
        ))
    catch
      case e: Exception =>
        System.err.println(s"  Warning: could not read $path: ${e.getMessage}")
        None

  // ── JSON summary ──────────────────────────────────────────────────────────

  private def writeSummaryJson(repo: String, scenario: String, entries: List[BenchmarkEntry], outputDir: os.Path): Unit =
    val out = outputDir / "summary.json"
    val json = ujson.Obj(
      "repo" -> ujson.Str(repo),
      "scenario" -> ujson.Str(scenario),
      "entries" -> ujson.Arr(entries.map { e =>
        ujson.Obj(
          "tool" -> ujson.Str(e.tool),
          "mean" -> ujson.Num(e.mean),
          "stddev" -> ujson.Num(e.stddev),
          "min" -> ujson.Num(e.min),
          "max" -> ujson.Num(e.max),
          "median" -> ujson.Num(e.median)
        )
      }: _*)
    )
    os.write.over(out, upickle.default.write(json, indent = 2))
    println(s"  Wrote $out")

  // ── Markdown summary ─────────────────────────────────────────────────────

  private def writeSummaryMd(repo: String, scenario: String, entries: List[BenchmarkEntry], outputDir: os.Path): Unit =
    val title = scenario.replace("-", " ").split(" ").map(_.capitalize).mkString(" ")
    val sb = new StringBuilder
    sb.append(s"# JVM Build Tools Benchmark — $repo — $title\n\n")
    sb.append("| Tool | Mean | Stddev | Min | Max |\n")
    sb.append("|------|-----:|-------:|----:|----:|\n")
    for e <- entries do
      sb.append(s"| ${e.tool} | ${fmtMs(e.mean)} | ± ${fmtMs(e.stddev)} | ${fmtMs(e.min)} | ${fmtMs(e.max)} |\n")
    sb.append("\n")
    val out = outputDir / "summary.md"
    os.write.over(out, sb.toString)
    println(s"  Wrote $out")

  // ── ASCII bar chart ───────────────────────────────────────────────────────

  private val BarWidth = 40
  private val BarChar  = "█"

  private def asciiBarChart(repo: String, scenario: String, entries: List[BenchmarkEntry]): Unit =
    if entries.isEmpty then return

    val maxMean = entries.map(_.mean).max
    val maxToolLen = entries.map(_.tool.length).max

    println(s"\n  [$repo] $scenario (mean time, lower is better)")
    for e <- entries do
      val barLen = if maxMean > 0 then Math.round((e.mean / maxMean) * BarWidth).toInt else 0
      val bar = BarChar * barLen
      val label = e.tool.padTo(maxToolLen, ' ')
      val meanMs = e.mean * 1000
      val stddevMs = e.stddev * 1000
      val barPadded = bar.padTo(BarWidth, ' ')
      println(f"    $label  $barPadded  $meanMs%8.1f ms ± $stddevMs%.1f")

  // ── SVG chart (JFreeChart + jfreesvg) ────────────────────────────────────

  private def writeSvgChart(repo: String, scenario: String, entries: List[BenchmarkEntry], outputDir: os.Path): Unit =
    try
      import org.jfree.chart.{ChartFactory, StandardChartTheme}
      import org.jfree.chart.plot.{CategoryPlot, PlotOrientation}
      import org.jfree.chart.renderer.category.BarRenderer
      import org.jfree.data.category.DefaultCategoryDataset

      // Build dataset — reversed so fastest is at top
      val reversed = entries.reverse
      val dataset = new DefaultCategoryDataset()
      for e <- reversed do
        dataset.addValue(e.mean * 1000, "Mean", e.tool)

      // Create chart
      val chart = ChartFactory.createBarChart(
        "",
        "",
        "Mean time (ms) — lower is better",
        dataset,
        PlotOrientation.HORIZONTAL,
        false,
        false,
        false
      )

      val title = s"$repo: ${scenario.replace("-", " ").split(" ").map(_.capitalize).mkString(" ")}"
      chart.setTitle(title)

      // Style
      val theme = StandardChartTheme.createJFreeTheme()
      theme.apply(chart)

      val plot = chart.getPlot.asInstanceOf[CategoryPlot]
      plot.setBackgroundPaint(java.awt.Color.WHITE)
      plot.setRangeGridlinePaint(java.awt.Color.LIGHT_GRAY)
      plot.setOutlineVisible(false)

      // Custom renderer with per-bar colors
      val renderer = new BarRenderer {
        override def getItemPaint(row: Int, column: Int): java.awt.Paint =
          val tool = reversed(column).tool
          try hexToColor(ToolColors.get(tool))
          catch case _: Exception => java.awt.Color.GRAY
      }
      renderer.setDrawBarOutline(false)
      plot.setRenderer(renderer)

      // Size
      val height = Math.max(150, entries.size * 40 + 100)
      val width = 800

      // Export to SVG via jfreesvg
      import org.jfree.graphics2d.svg.SVGGraphics2D
      val svgG2d = new SVGGraphics2D(width, height)
      val rect = new java.awt.Rectangle(0, 0, width, height)
      svgG2d.setClip(rect)
      chart.draw(svgG2d, rect)

      val svgText = svgG2d.getSVGElement
      val out = outputDir / "chart.svg"
      os.write.over(out, svgText)
      println(s"  Wrote $out")
    catch
      case e: Exception =>
        System.err.println(s"  SVG chart generation failed for $scenario: ${e.getMessage}")

  // ── HTML index page ───────────────────────────────────────────────────────

  private def writeIndexHtml(allData: Map[String, Map[String, List[BenchmarkEntry]]], outputDir: os.Path): Unit =
    val timestamp = ZonedDateTime.now(ZoneOffset.UTC).format(
      DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm 'UTC'")
    )

    val sections = new StringBuilder
    var first = true
    for (repo, scenarios) <- allData.toSeq.sortBy(_._1) do
      val openAttr = if first then " open" else ""
      first = false

      val charts = new StringBuilder
      for (scenario, _) <- scenarios.toSeq.sortBy(_._1) do
        val scenarioTitle = scenario.replace("-", " ").split(" ").map(_.capitalize).mkString(" ")
        charts.append(
          s"""      <figure>
             |        <img src="$repo/$scenario/chart.svg" alt="$scenario chart" loading="lazy">
             |        <figcaption>$scenarioTitle
             |          (<a href="$repo/$scenario/summary.json">json</a>,
             |           <a href="$repo/$scenario/summary.md">md</a>)
             |        </figcaption>
             |      </figure>
             |""".stripMargin
        )

      sections.append(
        s"""  <details$openAttr>
           |    <summary>$repo</summary>
           |    <div class="repo-body">
           |      <div class="charts">
           |$charts
           |      </div>
           |    </div>
           |  </details>
           |""".stripMargin
      )

    val html =
      s"""<!DOCTYPE html>
         |<html lang="en">
         |<head>
         |  <meta charset="UTF-8">
         |  <meta name="viewport" content="width=device-width, initial-scale=1.0">
         |  <title>JVM Build Tools Benchmark</title>
         |  <style>
         |  body { font-family: system-ui, sans-serif; max-width: 960px; margin: 2rem auto; padding: 0 1rem; color: #222; }
         |  h1 { font-size: 1.6rem; }
         |  details { border: 1px solid #ddd; border-radius: 6px; margin: 1rem 0; }
         |  summary { cursor: pointer; padding: 0.75rem 1rem; font-size: 1.1rem; font-weight: 600;
         |            background: #f5f5f5; border-radius: 6px; list-style: none; }
         |  summary::-webkit-details-marker { display: none; }
         |  summary::before { content: "▶ "; font-size: 0.8em; color: #666; }
         |  details[open] summary::before { content: "▼ "; }
         |  .repo-body { padding: 1rem; }
         |  .charts { display: flex; flex-wrap: wrap; gap: 1rem; }
         |  figure { margin: 0; flex: 1 1 400px; }
         |  figure img { width: 100%; height: auto; border: 1px solid #eee; border-radius: 4px; }
         |  figcaption { font-size: 0.85rem; color: #555; margin-top: 0.3rem; text-align: center; }
         |  footer { margin-top: 2rem; font-size: 0.8rem; color: #999; }
         |  </style>
         |</head>
         |<body>
         |  <h1>JVM Build Tools Benchmark</h1>
         |  <p>Compilation time comparison across build tools on real-world repos.
         |     Each chart shows mean time (lower is better).</p>
         |
         |$sections
         |
         |  <footer>Generated $timestamp</footer>
         |</body>
         |</html>
         |""".stripMargin

    val out = outputDir / "index.html"
    os.write.over(out, html)
    println(s"  Wrote $out")

  // ── Color utility ─────────────────────────────────────────────────────────

  private def hexToColor(hex: String): Color =
    val h = hex.stripPrefix("#")
    new Color(Integer.parseInt(h, 16))
